src/tools/rbd_mirror/ImageMap.cc

```diff
@@ -405,15 +405,16 @@ void ImageMap<I>::update_images_removed(
     bool image_mapped = (info.instance_id != image_map::UNMAPPED_INSTANCE_ID);
 
     bool image_removed = image_mapped;
-    bool peer_removed = false;
+    bool no_remote_peer = false;
     auto peer_it = m_peer_map.find(global_image_id);
     if (peer_it != m_peer_map.end()) {
       auto& peer_set = peer_it->second;
-      peer_removed = peer_set.erase(peer_uuid);
+      auto peer_removed = peer_set.erase(peer_uuid);
+      no_remote_peer = peer_removed && (peer_set.size() == peer_set.count(""));
       image_removed = peer_removed && peer_set.empty();
     }
 
-    if (image_mapped && peer_removed && !peer_uuid.empty()) {
+    if (image_mapped && no_remote_peer && !peer_uuid.empty()) {
```

Comment:
@trociny I have been debugging an issue where the image was not removed in the journal mode when I have 3 clusters. This code is perfectly fine (at least I think it is...) for the snapshot mode because of the following code:
https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc#L86-L89
https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc#L1474-L1489

So in the snapshot mode when the primary is removed it first go into `ImageMap<I>::update_images_removed` which clean the peer from the  m_peer_map and does nothing else (because there are some other peer still involved) but then the `handle_image_update_notify` is triggered and restart the Replayer which realize that the primary is missing and move the image to trash (which then also triggers `ImageMap<I>::update_images_removed`).

In the journal mode there is no `handle_image_update_notify` right now. So when there is 3 clusters involved `mageMap<I>::update_images_removed` is called and remove the peer, but because some peer remains nothing else is done and it stop there without stopping the replay and moving the image to the trash...

So now what should be the correct implementation? To me we could either:
- Try to handle that  in `ImageMap<I>::update_images_removed` so that any remote peer would trigger the image removal. But there are probably some corner case for instance removing a peer which has not the primary image but a secondary would maybe trigger this as well (not fully sure yet if this a problem or not though).
- Have a similar code than the snapshot replayer with the `librbd::UpdateWatchCtx ` and restart the replayer on `handle_notify`

I think if this doesn't do some weird behavior it should be the second option. Also as a bonus the behavior of both replayer would be more similar... WDYT?

Reply:
Right now the second option seems more preferable to me too. Lets try it?